### PR TITLE
actually check for module existence in manager

### PR DESCRIPTION
**Pull request by xylix**
Will improve module detection, and aims to add config file support for _available_modules and autostart lists

Merges and closes #27 and closes #48

TODOs:

 - [ ] Needs testing on Windows
 - [x] ~~Write config as commented out toml instead of old format~~ We're skipping this.
 - [x] Fix finding bundled executables (currently only system executables are found correctly

### CODE REVIEW

File: aw_qt/manager.py [hunk 1]
```diff
@@ -138,8 +138,10 @@ def __init__(self, testing: bool=False) -> None:
             # "aw-watcher-network"
         ]
 
-        # TODO: Filter away all modules not available on system
-        self.modules = {name: Module(name, testing=testing) for name in _possible_modules}
+        self.modules = []
+        for name in _possible_modules:
+            if not _locate_executable(name):
```
Comment (ErikBjare) [inline aw_qt/manager.py]: Maybe log an error if a module couldn't be located?

Comment (ErikBjare) [reply aw_qt/manager.py]: Travis fails on type checking, it seems that the code expects `self.modules` to be a dict.

**Comment by ErikBjare** [comment]
If you want to further improve `manager.py` take a look at https://github.com/ActivityWatch/aw-qt/pull/27 and https://github.com/ActivityWatch/aw-qt/pull/48.

Would be nice to finally get them merged or closed.

**Comment by xylix** [comment]
Yeah I'll try to get those done.

**Comment by xylix** [comment]
I added the https://github.com/ActivityWatch/aw-qt/pull/48 changes and will look at the https://github.com/ActivityWatch/aw-qt/pull/27 changes next. 

#27 has some generic improvements which I'll at least integrate, but having to install all modules into path would be something I wouldn't prefer.


I think it'd be better to have a global config file with a list which just happens to include all the aw's own modules by default. This'd make it much easier to add external modules without changing the source code and would also theoretically support having a "/modules" directory somewhere which could be sourced by the user.

@ErikBjare Whats your opinion about automatic path detection vs a "possible modules" list and trying to find those modules? It's been quite a few years since #27

**Comment by ErikBjare** [comment]
About the config introduced in #27, I'd like to see the config file use `toml` instead of the config format used by `aw_core.load_config`. The `aw_core.load_config` also incorrectly writes defaults to file. Commented out defaults can be written to file if none exists, to clarify which options exist.

An option is to simply rewrite `aw_core.load_config` to use `toml` and use the "write commented out defaults if no config exist" rule, but that might require writing some type of migration from existing configs.

**Comment by ErikBjare** [comment]
I'd like @johan-bjareholt's comments on that as well.

**Comment by ErikBjare** [comment]
>Whats your opinion about automatic path detection vs a "possible modules" list and trying to find those modules?

So I guess there are two questions:

 - **What** do we look for when looking for modules?
   - Config vs all executables with a `aw-` prefix
   - We currently only look for the modules in the hard-coded list
 - **Where** do we look for module executables?
   - Special folders (such as subfolders of the aw-qt executable) and/or PATH
   - We currently look in the directory of the aw-qt executable (this might be redundant with how packaging now) as well as subfolders starting with `aw-` (I haven't verified this in the code, but that's what I remember)
   - If two modules with the same name is found (for example, aw-server is found both in a subdirectory and PATH) then a module executable found in the subdirectory should take precedence. You can also have a module twice in your PATH, but there the normal PATH precedence rules should apply (i.e. always pick the first found).

It doesn't make much sense to me to hard-code possible modules (as we will be able to figure out which are available anyway from the PATH and subdirectories) but there needs to be a way to specify which modules to run on startup (apart from the CLI flag).

**Comment by xylix** [comment]
@ErikBjare I have now merged #27 and #48, so they can be closed. I also generally improved typing support everywhere around aw-qt.

Still TODO before merging this: 
- [ ] Write config as commented out toml instead of old format
- [ ] Fix finding bundled executables (currently only system executables are found correctly

**Comment by xylix** [comment]
@ErikBjare Looking into writing out the commented out TOML, I realised it requires me adding support to aw-core and changing the default behaviour of the "load_config" function. Currently aw-core/aw_core/config.py's load config function overwrites the old config on every load, with the loaded one, also adding all new options at the same time.

So the config changes will require changing aw-core's default behaviour, and we'll need to validate the change won't break any current modules. 

I personally think the most sane way of implementing this would be to finish this PR with the current config behaviour and migrate aw-qt's config file behaviour along with all other modules. The config change will need it's own aw-core PR anyway. I'd also like to investigate exporting aw-server-rust's config reading / writing behaviour for use in aw-core. @johan-bjareholt Do you think that could be possible?

I think next steps will be me fixing the bundled module detection behaviour, removing the redundant "possible modules" behaviour and then starting work on the aw-core PR. I don't mind us waiting on merging this until the changes on aw-core side config system are finished.

**Comment by ErikBjare** [comment]
@xylix I started working on TOML support in aw-core here: https://github.com/ActivityWatch/aw-core/pull/88

File: aw_qt/manager.py [hunk 2]
```diff
@@ -1,68 +1,128 @@
...
-        # TODO: Actually check if it is in PATH
-        logger.debug("Trying to start {} using PATH (executable not found in: {})"
-                     .format(name, exec_paths))
-        return [name]
+        logger.info("Found no bundles modules")
+    return modules
+
+""" Find all aw- modules in PATH """
+def _discover_modules_system() -> List[str]:
+    search_paths = os.environ["PATH"].split(":")
```
Comment (RoBobik) [inline aw_qt/manager.py:81]: Hi, I haven't tested these changes myself but I think this line won't work on Windows, as Windows uses semicolon (`;`) to separate PATH items. Python has `os.pathsep` which returns the correct separator, so you should be able to do:
```
search_paths = os.environ["PATH"].split(os.pathsep)
```

Comment (RoBobik) [reply aw_qt/manager.py:81]: Actually, [get_exec_path()](https://docs.python.org/3.7/library/os.html#os.get_exec_path) might be a cleaner solution (for Python 3.2+):
```
search_paths = os.get_exec_path()
```

Credits to @martincizek.

Comment (ErikBjare) [reply aw_qt/manager.py:81]: @RoBobik great points, needs fixing before merge.

Comment (xylix) [reply aw_qt/manager.py]: Yeah good points. Changed to using os.get_exec_path() in 68314f0

File: aw_qt/manager.py [hunk 3]
```diff
@@ -87,7 +145,7 @@ def stop(self) -> None:
             logger.warning("Tried to stop module {}, but it wasn't running".format(self.name))
         else:
             if not self._process:
-                logger.error("No reference to process object")
+                logger.error("")
```
Comment (ErikBjare) [inline aw_qt/manager.py:148]: What happened here?

File: aw_qt/manager.py [hunk 4]
```diff
@@ -71,8 +131,6 @@ def start(self) -> None:
         # There is a very good reason stdout and stderr is not PIPE here
         # See: https://github.com/ActivityWatch/aw-server/issues/27
         self._process = subprocess.Popen(exec_cmd, universal_newlines=True, startupinfo=startupinfo)
-
-        # Should be True if module is supposed to be running, else False
```
Comment (ErikBjare) [inline aw_qt/manager.py:75]: Why remove this comment?

Comment (ErikBjare) [reply aw_qt/manager.py:75]: Nvm, noticed now it was moved up.

**Comment by ErikBjare** [comment]
I merged master into the branch and fixed a few things. Should be ready to merge now.

**Comment by ErikBjare** [comment]
I'm just going to YOLO this and assume it will work fine on Windows (my only concern is the use of `os.access`, but I think it'll work). 

Merging as soon as CI passes.

**Comment by xylix** [comment]
Great, nice that you got this finished!

**Comment by Alwinator** [comment]
I will test it for Windows. If I find a error, I will report it here.

**Pull request merged**